File: o2a/quota/adapters/declarative.py
"""declarative 适配器：用 accounts[].quota 声明式窗口，接入套餐目录。

quota_source="declarative" 时使用该适配器：
- accounts[].quota = {"plan": "glm-coding-plan"} 或
  {"windows": [{"kind":"month","unit":"requests","limit":200}], "plan": "my-plan"}
- 需要本地用量时通过 stats_dir 聚合（与 manual 一致）
- 不依赖外网，适合 Sub2API/New API 等“声明额度”场景
"""

from ..base import QuotaAdapter, QuotaContext, empty_window, make_snapshot, window_start
from ._stats_util import count_requests, count_tokens


class DeclarativeQuotaAdapter(QuotaAdapter):
    name = "declarative"
    source = "plan_config"

    def fetch(self, ctx: QuotaContext):
        acc = ctx.account
        if acc is None:
            return None
        cfg = acc.quota or {}
        plan_name = cfg.get("plan")
        windows_cfg = cfg.get("windows")
        if not plan_name and not windows_cfg:
            return None
        windows = []
        for w in windows_cfg or []:
            if not isinstance(w, dict):
                continue
            kind = w.get("kind", w.get("period", "month"))
            unit = w.get("unit", "requests")
            limit = w.get("limit")
            period = w.get("period", "month")
            if period not in ("day", "week", "month"):
                period = "month"
            start = window_start(ctx.now(), period)
            used = count_requests(ctx.stats_dir, acc.id, start) if unit == "requests" else \
                count_tokens(ctx.stats_dir, acc.id, start)
            windows.append(empty_window(kind, unit, used, limit,
                                        reset_at=ctx.iso(window_start(ctx.now(), "day"))))
        plan = None
        if plan_name:
            plan = {"name": plan_name, "included": cfg.get("included") or {},
                    "overage": cfg.get("overage") or {}, "free_tier": cfg.get("free_tier") or {}}
        return make_snapshot(self.name, windows or [], source=self.source,
                             plan=plan, now_fn=ctx.now_fn)

File: crates/o2a-quota/src/adapters/declarative.rs
//! declarative 适配器：用 accounts[].quota 声明式窗口，接入套餐目录。
//!
//! accounts[].quota = {"plan": "glm-coding-plan"} 或
//! {"windows": [{"kind":"month","unit":"requests","limit":200}], "plan": "my-plan"}
//! 需要本地用量时通过 stats_dir 聚合（与 manual 一致）；不依赖外网。

use serde_json::{json, Map, Value};

use crate::base::{empty_window, make_snapshot, window_start, QuotaContext, QuotaError};
use crate::registry::QuotaAdapter;
use crate::stats_util::{count_requests, count_tokens};

pub struct DeclarativeQuotaAdapter;

#[async_trait::async_trait]
impl QuotaAdapter for DeclarativeQuotaAdapter {
    fn name(&self) -> &'static str {
        "declarative"
    }

    fn source(&self) -> &'static str {
        "plan_config"
    }

    async fn fetch(&self, ctx: &QuotaContext) -> Result<Option<Value>, QuotaError> {
        let cfg = ctx.account.quota.clone().unwrap_or_default();
        let plan_name = cfg
            .get("plan")
            .and_then(Value::as_str)
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string());
        let windows_cfg = cfg.get("windows").and_then(Value::as_array).cloned();
        if plan_name.is_none() && windows_cfg.is_none() {
            return Ok(None);
        }
        let now = ctx.now();
        let mut windows = Vec::new();
        for w in windows_cfg.unwrap_or_default() {
            let Some(w) = w.as_object() else {
                continue;
            };
            let kind = w
                .get("kind")
                .and_then(Value::as_str)
                .or_else(|| w.get("period").and_then(Value::as_str))
                .unwrap_or("month")
                .to_string();
            let unit = w
                .get("unit")
                .and_then(Value::as_str)
                .unwrap_or("requests")
                .to_string();
            let limit = w.get("limit").cloned().unwrap_or(Value::Null).as_f64();
            let mut period = w
                .get("period")
                .and_then(Value::as_str)
                .unwrap_or("month")
                .to_string();
            if !["day", "week", "month"].contains(&period.as_str()) {
                period = "month".into();
            }
            let start = window_start(&now, &period);
            let used = if unit == "requests" {
                count_requests(&ctx.stats_dir, &ctx.account.id, &start) as f64
            } else {
                count_tokens(&ctx.stats_dir, &ctx.account.id, &start) as f64
            };
            windows.push(empty_window(
                &kind,
                &unit,
                used,
                limit,
                Some(ctx.iso(&window_start(&now, "day"))),
            ));
        }
        let plan = plan_name.map(|name| {
            let empty = Map::new();
            json!({
                "name": name,
                "included": cfg.get("included").cloned().unwrap_or(Value::Object(empty.clone())),
                "overage": cfg.get("overage").cloned().unwrap_or(Value::Object(empty.clone())),
                "free_tier": cfg.get("free_tier").cloned().unwrap_or(Value::Object(empty)),
            })
        });
        let snap = make_snapshot(self.name(), windows, self.source(), plan, false, &now);
        Ok(Some(snap))
    }
}
